src/networking.c

```diff
@@ -1911,31 +1936,75 @@ client *lookupClientByID(uint64_t id) {
 void writeToReplica(client *c) {
     /* Can be called from main-thread only as replica write offload is not supported yet */
     serverAssert(inMainThread());
-    int nwritten = 0;
+    ssize_t nwritten = 0;
+    zeroCopyRecord *ongoing_zero_copy_write = NULL;
     serverAssert(c->bufpos == 0 && listLength(c->reply) == 0);
     while (clientHasPendingReplies(c)) {
         replBufBlock *o = listNodeValue(c->repl_data->ref_repl_buf_node);
         serverAssert(o->used >= c->repl_data->ref_block_pos);
 
         /* Send current block if it is not fully sent. */
         if (o->used > c->repl_data->ref_block_pos) {
-            nwritten = connWrite(c->conn, o->buf + c->repl_data->ref_block_pos, o->used - c->repl_data->ref_block_pos);
+            size_t data_len = o->used - c->repl_data->ref_block_pos;
+            int use_zerocopy = shouldUseZeroCopy(c->conn, data_len);
+            if (use_zerocopy) {
+                /* Lazily enable zero copy at the socket level only on first use */
+                if (!c->zero_copy_tracker) {
+                    connSetZeroCopy(c->conn, 1);
+                    c->zero_copy_tracker = createZeroCopyTracker();
+                }
+                nwritten = zeroCopyWriteToConn(c->conn, o->buf + c->repl_data->ref_block_pos, data_len);
+            } else {
+                nwritten = connWrite(c->conn, o->buf + c->repl_data->ref_block_pos, data_len);
+            }
```

Comment:
This seems like it could be encapsulated in the connection abstraction maybe? I wonder if the tracker should be a client or a connection owned? I think it makes more sense that it will be part of the connection.